micropip/_micropip.py

```diff
@@ -692,3 +695,72 @@ def _list():
             source_ = pkg_source
         packages[name] = PackageMetadata(name=name, version=version, source=source_)
     return packages
+
+
+def add_mock_package(name, version, *, modules=None):
+    """
+       Add a mock version of a package to the package dictionary.
+
+       This means that if it is a dependency, it is skipped on install.
+
+       By default a single empty module is installed with the same
+       name as the package. You can alternatively give one or more modules to make a
+       set of named modules.
+
+       The modules parameter is a dictionary mapping module name to module text.
+
+        e.g.
+       ``
+       {
+          mylovely_module:'''
+            def module_method(an_argument):
+                print("This becomes a module level argument")
+            module_value = "this value becomes a module level variable"
+
+            print("This is run on import of module")
+          '''
+       }
+
+
+       ``
+
+    Parameters
+    ----------
+    name : ``str ``
+
+        Package name to add
+
+    modules : ``Optional(dict) ``
+
+        Dictionary of module_name:string pairs.
+        The string contains the source of the mock module or is blank for
+        an empty module.
+    """
+    if modules is None:
+        # make a single mock module with this name
+        modules = {name: ""}
+    # make empty mock modules with the requested names in user site packages
+    site_packages = Path(site.getusersitepackages())
+    metadata_file = site_packages / (name + "-" + version + ".egg-info")
+    with open(metadata_file, "w") as mf:
+
+        mf.write(
+            f"""Metadata-Version: 1.1
+Name: {name}
+Version: {version}
+Summary: {name} mock module generated by micropip
+Author-email: {name}@micro.pip.non-working-fake-host
+"""
+        )
+        for n, s in modules.items():
+            mf.write(f"Provides:{n}\n")
+
+            if s is None:
+                s = ""
+            s = dedent(s)
+            path_parts = n.split(".")
+            dir_path = Path(site_packages, *path_parts)
+            os.makedirs(dir_path, exist_ok=True)
+            init_file = dir_path / "__init__.py"
```

Comment:
Do we actually need to create these physical files?

I was thinking we could maybe directly initialize `sys.modules` with the mapping from `modules: dict[str, ModuleType | str]`? That way it's also easy to use `MagicMock` as a module or any other custom object defined at runtime to make a given dependency work.

 Though yes, the API to create modules dynamically via [`importlib.util.module_from_spec`](https://docs.python.org/3/library/importlib.html#importlib.util.module_from_spec)  or by [inheriting from ModuleType](https://stackoverflow.com/a/67813054/1791279) isn't very straightforward.

Reply:
Maybe making another custom Finder/Loader would be a good way to go. The API for that isn't too bad.

Reply:
I actually wrote it initially using mock objects and putting them into sys.modules, but I reverted to writing them out to files. I did it like that because it is really simple and reliable, things like file paths etc. just work. It also works nicely in terms of loading modules, in that init code is only run on actual import. Oh and in the case that you're running somewhere with persistent file system, your mock modules continue to exist on reload of python.

It's a pretty minor change - you just create a module and exec the code into it in the loader, I can put that back in if you reckon the possibility of using mock objects makes sense.

Reply:
Actually I think rather than take an object it needs to take a callable which takes a module object and does whatever needs to be done on it to make it. I.e. the equivalent of loader.exec_module

That way you can mock module initialisation correctly.

Reply:
One further thought on this - at some point I want to implement some kind of mega wheel option, to bundle a set of wheels and all deps into a single wheel with no deps. For that use case keeping persistent mock modules would be nice. If it's okay I'd like to keep that as an option in micropip?

It would also be nice for any point that pyodide is run in a persistent file system. 

I'm thinking that a flag "persistent" could be added (which only works for string modules) and then the behaviour with finder and loader could be used in the case that the flag isn't set.